sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DiagnosticsProvider.java

```diff
@@ -0,0 +1,1338 @@
+        Object cosmosCtx = traceContext.getData(COSMOS_DIAGNOSTICS_CONTEXT_KEY).orElse(null);
+
+        if (cosmosCtx instanceof CosmosDiagnosticsContext) {
+            return (CosmosDiagnosticsContext) cosmosCtx;
+        }
+
+        throw new IllegalStateException("CosmosDiagnosticsContext not present.");
+    }
+
+    /**
+     * Stores {@link Context} in Reactor {@link reactor.util.context.Context}.
+     *
+     * @param traceContext {@link Context} context with trace context to store.
+     * @return {@link reactor.util.context.Context} Reactor context with trace context.
+     */
+    public static reactor.util.context.Context setContextInReactor(Context traceContext) {
+        return reactor.util.context.Context.of(REACTOR_TRACING_CONTEXT_KEY, traceContext);
+    }
+
+    /**
+     * For each tracer plugged into the SDK a new tracing span is created.
+     * <p>
+     * The {@code context} will be checked for containing information about a parent span. If a parent span is found the
+     * new span will be added as a child, otherwise the span will be created and added to the context and any downstream
+     * start calls will use the created span as the parent.
+     *
+     * @param context Additional metadata that is passed through the call stack.
+     * @return An updated context object.
+     */
+    public Context startSpan(
+        String spanName,
+        CosmosDiagnosticsContext cosmosCtx,
+        Context context) {
+
+        checkNotNull(spanName, "Argument 'spanName' must not be null.");
+        checkNotNull(cosmosCtx, "Argument 'cosmosCtx' must not be null.");
+
+        ctxAccessor.startOperation(cosmosCtx);
+        Context local = Objects
+            .requireNonNull(context, "'context' cannot be null.")
+            .addData(COSMOS_DIAGNOSTICS_CONTEXT_KEY, cosmosCtx);
+
+        if (this.cosmosTracer == null) {
+            return local;
+        }
+
+        return this.cosmosTracer.startSpan(spanName, cosmosCtx, local);
+    }
+
+    /**
+     * Given a context containing the current tracing span the span is marked completed with status info from
+     * {@link Signal}.  For each tracer plugged into the SDK the current tracing span is marked as completed.
+     *
+     * @param signal  The signal indicates the status and contains the metadata we need to end the tracing span.
+     */
+
+    public <T> void endSpan(
+        Signal<T> signal,
+        int statusCode,
+        Integer actualItemCount,
+        Double requestCharge,
+        CosmosDiagnostics diagnostics
+    ) {
+        // called in PagedFlux - needs to be exception less - otherwise will result in hanging Flux.
+        try {
+            this.endSpanCore(signal, statusCode, actualItemCount, requestCharge, diagnostics);
+        } catch (Throwable error) {
+            LOGGER.error("Unexpected exception in DiagnosticsProvider.endSpan. ", error);
+            System.exit(9901);
```

Comment:
System.exit (here and below)?
Is this really needed and there is no other option other then exiting the VM?

Reply:
I added this here for now to find possible issues during testing this week. If any exception happens here the Reactor Flux/Mono is just stuck and will never complete - that obviously is a pretty bad issue.